### PR TITLE
Add bladeswap trades

### PR DESCRIPTION
**Pull request by Hosuke**
#6271 

### Contribution type
Please check the type of contribution this pull request is for:

- [x] New spell(s)
- [ ] Adding to existing spell lineage
- [ ] Bug fix

**Note:** You can safely discard any section below which doesn't apply based on selection above

---

### For new spell(s)
If you are building new spell(s), please provide the following information:

- **Spell name(s):** *schema.table_name*
- **Description:** [Detailed description of the new spell(s) and their purpose]
- **Who are the new spell(s) for?** [Internal team or general community]
- **How will the new spell(s) be used downstream?** [Description of downstream usage]
- **Implementation details:** [Information on how the spell(s) are implemented]
- **Test instructions:** [How to test the new spell(s)]
- **Related issue(s):** [Link to related issues, if any]

---

### For adding to existing spell lineage
If you are adding to an existing spell lineage, please provide the following information:

- **Description:** [Description of the changes made]

---

### For bug fixes
If you are fixing a bug, please provide the following information:

- **Description:** [Description of the bug fix]
- **Steps to reproduce:** [How to reproduce the bug]
- **Implementation details:** [Information on how the bug was fixed]
- **Test instructions:** [How to test the fix]
- **Related issue(s):** [Link to related issues, if any]

---

### Additional information
Please provide any additional information that might help us review your pull request:

- [Any additional information]

---

Thank you for your contribution!

### CODE REVIEW

**Comment by jeff-dude** [comment]
i see this one is still marked as `wip` -- let me know the status here 🙏 
also some merge conflicts to resolve

File: sources/_sector/dex/trades/blast/_sources.yml [hunk 1]
```diff
@@ -17,3 +17,8 @@ sources:
       - name: ThrusterFactory_Point_3_Fee_evt_PairCreated
       - name: ThrusterPair_V2_1_Fee_evt_Swap
       - name: ThrusterFactory_1_Fee_evt_PairCreated
+  - name: bladeswap_blast
+    tables:
+      - name: XYKPoolFactory_V2_evt_PoolCreated
+      - name: XYKPool_evt_Swap
+      - name: Vault_Router_evt_Swap
```
Comment (Hosuke) [inline sources/_sector/dex/trades/blast/_sources.yml:24]: The swap events here may not be very clear:
https://dune.com/queries/3925455

here is an example event log:
https://blastscan.io/tx/0x3f9a53ffe1369136af86c1fff61868ede338be17ace446b8880249c371379d10#eventlog

**Comment by Hosuke** [comment]
I just submit contract `0x9b6D09975E29D1888b98B83e31e72c00bC4D93C5` for decoding.
We may utilize the new contract then.
https://bladeswap.gitbook.io/bladeswap/resources/contract-addresses

File: sources/_sector/dex/trades/blast/_sources.yml (hunk 1 above)
Comment (Hosuke) [reply sources/_sector/dex/trades/blast/_sources.yml:24]: Sample bladeswap_blast.base_trades based on manually decoding:
https://dune.com/queries/3925627

File: dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+WITH source_expanded AS (
+    SELECT
+        delta[1] as amount0,
+        delta[2] as amount1,
+        bytearray_substring (tokenRef[1], 13, 20) as token0,
+        bytearray_substring (tokenRef[2], 13, 20) as token1,
+        *
+    FROM
+        {{ source('bladeswap_blast', 'Vault_Router_evt_Swap') }}
+    WHERE CARDINALITY(delta) > 1
```
Comment (Hosuke) [inline dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql:22]:
```suggestion
    WHERE CARDINALITY(delta) = 2
```

File: sources/_sector/dex/trades/blast/_sources.yml (hunk 1 above)
Comment (Hosuke) [reply sources/_sector/dex/trades/blast/_sources.yml:24]: However multiple tokens (>2) swapping may happen here.

**Comment by jeff-dude** [comment]
doing some (quick) digging around. looks like bladeswap on blast is a [fork of velocore](https://defillama.com/protocol/bladeswap-amm#information). according to velocore [docs](https://docs.velocore.xyz/dex-innovation/token-vault), dex is inspired by balancer vaults. i wonder if we can use the balancer v2 code (from balancer macro) as inspiration to properly pull this data?

**Comment by Hosuke** [comment]
I will try to utilize the `ROW_NUMBER()` part, besides the table schema looks different at the moment.

File: dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql [hunk 3]
```diff
@@ -10,21 +10,32 @@
...
         *
     FROM
-        {{ source('bladeswap_blast', 'Vault_Router_evt_Swap') }}
-    WHERE CARDINALITY(delta) = 2
-    {% if is_incremental() %}
-      AND {{ incremental_predicate('evt_block_time') }}
-    {% endif %}
-
+        filtered_source
+--     WHERE CARDINALITY(delta) = 2
```
Comment (Hosuke) [inline dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql:36]:
```suggestion
     WHERE CARDINALITY(delta) = 2
```

File: dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+, filtered_source_expanded AS (
+    SELECT
+        delta[1] as amount0,
+        delta[2] as amount1,
+        bytearray_substring (tokenRef[1], 13, 20) as token0,
+        bytearray_substring (tokenRef[2], 13, 20) as token1,
+        *
+    FROM
+        filtered_source
+     WHERE CARDINALITY(delta) = 2
```
Comment (Hosuke) [inline dex/models/trades/blast/platforms/bladeswap_blast_base_trades.sql:36]:
```suggestion
    WHERE CARDINALITY(delta) = 2
```

**Comment by jeff-dude** [comment]
i noticed you left out the seed and commented out the test. is that cause we aren't certain the code is correct? or its difficult to navigate the explorer(s) to find tx's to test against?

**Comment by Hosuke** [comment]
> i noticed you left out the seed and commented out the test. is that cause we aren't certain the code is correct? or its difficult to navigate the explorer(s) to find tx's to test against?

You are right, I am not very confident with the current implementation (and the seed).
I just added them back.

File: dex/seeds/trades/bladeswap_blast_base_trades_seed.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+blockchain,project,version,block_date,tx_hash,evt_index,token_bought_address,token_sold_address,block_number,token_bought_amount_raw,token_sold_amount_raw
+blast,bladeswap,1,2024-07-22,0x404b62b3acf49cb54739538f6e0d1a71ba68561872f5ee655f364f03c14b9a95,41,0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee,0xd1fedd031b92f50a50c05e2c45af1adb4cea82f4,6415903,28706844152887539,6489521714789132087008
```
Comment (Hosuke) [inline dex/seeds/trades/bladeswap_blast_base_trades_seed.csv:2]: https://blastscan.io/tx/0x404b62b3acf49cb54739538f6e0d1a71ba68561872f5ee655f364f03c14b9a95

File: dex/seeds/trades/bladeswap_blast_base_trades_seed.csv [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+blast,bladeswap,1,2024-07-21,0x01996d5ae5da567e1765b65a03172135b597220f90816a0526961d2c2a6e3631,168,0xd1fedd031b92f50a50c05e2c45af1adb4cea82f4,0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee,6360267,243806794235376124005,1110890207385634
```
Comment (Hosuke) [inline dex/seeds/trades/bladeswap_blast_base_trades_seed.csv:3]: https://blastscan.io/tx/0x01996d5ae5da567e1765b65a03172135b597220f90816a0526961d2c2a6e3631

File: sources/_sector/dex/trades/blast/_sources.yml [hunk 7]
```diff
@@ -17,6 +17,12 @@ sources:
       - name: ThrusterFactory_Point_3_Fee_evt_PairCreated
       - name: ThrusterPair_V2_1_Fee_evt_Swap
       - name: ThrusterFactory_1_Fee_evt_PairCreated
+  - name: bladeswap_blast
```
Comment (Hosuke) [inline sources/_sector/dex/trades/blast/_sources.yml:20]: With extra sources introduced (which I may miss), we may implement the `bladeswap.trades` in a more proper way.

**Comment by jeff-dude** [comment]
as discussed in meeting, let's close this one for now. as always, keep the branch in case. i will communicate the complexity on this project.

**Pull request closed without merging**